File: sdk/swift-holons/Tests/HolonsTests/ResolveTests.swift
import XCTest

@testable import Holons

final class ResolveContractTests: XCTestCase {
  func testKnownSlug() throws {
    let sandbox = try UniformSandbox(prefix: "resolve-known")
    defer { try? FileManager.default.removeItem(at: sandbox.root) }

    try writeUniformPackage(
      at: sandbox.cwdPackageRoot.appendingPathComponent("resolve-known.holon", isDirectory: true),
      slug: "resolve-known",
      uuid: "uuid-resolve-known"
    )

    let context = makeResolveContext(sandbox)
    defer { context.restore() }

    let result = resolve(
      scope: LOCAL,
      expression: "resolve-known",
      root: sandbox.cwdPackageRoot.path,
      specifiers: CWD,
      timeout: NO_TIMEOUT
    )

    XCTAssertNil(result.error)
    XCTAssertEqual(result.ref?.info?.slug, "resolve-known")
  }

  func testMissingTarget() throws {
    let sandbox = try UniformSandbox(prefix: "resolve-missing")
    defer { try? FileManager.default.removeItem(at: sandbox.root) }

    let context = makeResolveContext(sandbox)
    defer { context.restore() }

    let result = resolve(
      scope: LOCAL,
      expression: "missing-target",
      root: sandbox.cwdPackageRoot.path,
      specifiers: CWD,
      timeout: NO_TIMEOUT
    )

    XCTAssertNil(result.ref)
    XCTAssertNotNil(result.error)
  }

  func testInvalidSpecifiers() throws {
    let sandbox = try UniformSandbox(prefix: "resolve-invalid-specifiers")
    defer { try? FileManager.default.removeItem(at: sandbox.root) }

    let context = makeResolveContext(sandbox)
    defer { context.restore() }

    let result = resolve(
      scope: LOCAL,
      expression: "anything",
      root: sandbox.cwdPackageRoot.path,
      specifiers: 0x40,
      timeout: NO_TIMEOUT
    )

    XCTAssertNotNil(result.error)
  }
}

private func makeResolveContext(_ sandbox: UniformSandbox) -> UniformDiscoveryContext {
  try? FileManager.default.createDirectory(at: sandbox.cwdPackageRoot, withIntermediateDirectories: true)
  return UniformDiscoveryContext(
    root: sandbox.cwdPackageRoot,
    opHome: sandbox.opHome,
    opBin: sandbox.opBin
  )
}

File: sdk/csharp-holons/Holons.Tests/ResolveTests.cs
namespace Holons.Tests;

public class ResolveTests
{
    [Fact]
    public void KnownSlug()
    {
        using var env = new RuntimeEnvironment();
        var packageDir = Path.Combine(env.Root, "known.holon");
        DiscoveryTestSupport.WritePackageHolon(
            packageDir,
            new PackageSeed("known", "uuid-known", "Known", "Slug"));

        var result = Discovery.Resolve(Discovery.LOCAL, "known", env.Root, Discovery.CWD, Discovery.NO_TIMEOUT);

        Assert.Null(result.Error);
        Assert.NotNull(result.Ref);
        Assert.Equal("known", result.Ref!.Info?.Slug);
        Assert.Equal(DiscoveryTestSupport.FileUrl(packageDir), result.Ref.Url);
    }

    [Fact]
    public void MissingTarget()
    {
        using var env = new RuntimeEnvironment();

        var result = Discovery.Resolve(Discovery.LOCAL, "missing", env.Root, Discovery.CWD, Discovery.NO_TIMEOUT);

        Assert.NotNull(result.Error);
        Assert.Null(result.Ref);
    }

    [Fact]
    public void InvalidSpecifiers()
    {
        using var env = new RuntimeEnvironment();

        var result = Discovery.Resolve(Discovery.LOCAL, "missing", env.Root, 0xFF, Discovery.NO_TIMEOUT);

        Assert.NotNull(result.Error);
        Assert.Null(result.Ref);
    }
}
